pkg/gatewayserver/gatewayserver.go

```diff
@@ -412,10 +419,23 @@ func (gs *GatewayServer) Connect(ctx context.Context, frontend io.Frontend, ids
 	if err != nil {
 		return nil, err
 	}
-	gs.connections.Store(uid, conn)
+	connEntry := connectionEntry{
+		Connection:   conn,
+		upstreamDone: make(chan struct{}),
+	}
+	for existing, exists := gs.connections.LoadOrStore(uid, connEntry); exists; existing, exists = gs.connections.LoadOrStore(uid, connEntry) {
```

Comment:
I see a potential deadlock here. 
- There's Conn A.
- Conn B now replaces Conn A and blocks until A is closed.
- Now Conn C will replace Conn B and force it to close.
- But if Conn B continues the loop before Conn C closes it  via `L431`, then Conn B again replaces Conn C and try to close it.

It's kinda rare but theoretically possible. How about disallowing a second (or multiple) connection and also closing the existing connection?
Then when the gateway retries we can have a clean connect from the beginning?

Reply:
>But if Conn B continues the loop before Conn C closes it 

This doesn't happen. `LoadOrStore()` is atomic; if it gets loaded, that existing connection gets closed and we can wait for that. If it doesn't get loaded, the connection is by definition stored, the loop breaks and the upstream handler gets started. Even if the upstream handler is not up yet, connection C can safely wait for the channel close, as connection B will definitely start that goroutine.